### ISSUE TITLE
embedding markdown package in app

### ISSUE THREAD

**Issue by typemytype** [Python-Markdown/markdown#961]
Inside an macOS app a package is not installed or pipped but used as plain package inside a `python<version>` folder inside `<appName>.app/Contents/Resources`.

the latest version (3.2.2) raises an error on import of `importlib_metadata` as it cannot find his own version... (traceback below), I assume, even if this could be fixed, `importlib_metadata` will not find any registered extensions as its will never will be installed and registers as an entry_point... 

Any advice on how to use markdown without installing it? this seems now not possible

thanks for all the good work!! 

```
File "/Contents/Resources/lib/python3.7/markdown/__init__.py", line 29, in <module>
  from .core import Markdown, markdown, markdownFromFile # noqa: E402
 File "/Contents/Resources/lib/python3.7/markdown/core.py", line 26, in <module>
  from . import util
 File "/Contents/Resources/lib/python3.7/markdown/util.py", line 34, in <module>
  import importlib_metadata as metadata
 File "/Contents/Resources/lib/python3.7/importlib_metadata/__init__.py", line 602, in <module>
  __version__ = version(__name__)
 File "/Contents/Resources/lib/python3.7/importlib_metadata/__init__.py", line 564, in version
  return distribution(distribution_name).version
 File "/Contents/Resources/lib/python3.7/importlib_metadata/__init__.py", line 537, in distribution
  return Distribution.from_name(distribution_name)
 File "/Contents/Resources/lib/python3.7/importlib_metadata/__init__.py", line 205, in from_name
  raise PackageNotFoundError(name)
importlib_metadata.PackageNotFoundError: importlib_metadata
```

**Comment by typemytype** [comment]
thanks for the requirement naming typo... but this does not fix the issue when the package is not installed...

you can safely assume when the import of `import importlib_metadata as metadata` fails`INSTALLED_EXTENSIONS` can be a empty set

https://github.com/Python-Markdown/markdown/blob/master/markdown/util.py#L85

It would be very handy to have markdown independent of entry-points, as the embedded extensions are accessible by direct import!

thanks

**Comment by mitya57** [comment]
But in your case it is installed, right? Otherwise `/Contents/Resources/lib/python3.7/importlib_metadata/` would not exist…

Anyway that package is part of Python 3.8 standard library, and I don't think there are many users who have older Python *and* have problems with installing such dependencies.

**Comment by typemytype** [comment]
the module `importlib_metadata` is embedded into the app, but fails on getting his own version is its is not installed with `pip` of `setup.py install`. This also fails with the embedded packages on 3.8. 

Apps are build with py2app and modules are collected and copied inside the app. There is no install, no setup.py, for both `markdown` as for `importlib_metadata`.

**Comment by waylan** [comment]
It looks like this was autoclosed when I merged #962. @typemytype feel free to reopen or leave a comment if the issue is not resolved.

Regarding entry points, the docs list three ways to enable an [extension][1]:

1. An instance of an extension (the preferred method). Import the class and pass in an instance.
2. Use a string containing the entry point name assigned to an extension.
3. Use a string containing Python dot notation (`'path.to.module:MyExtClass'`). For any extension which has a `makeExtension` function defined (all of the included extensions do), then the classname is optional (`'path.to.module'`).

If you are using the library in a setting where entrypoints don't work, then the other two options are still available to you. In fact option 3 was the preferred method throughout version 2.x. But users liked the short names from version 1.x so we brought them back as entry points.

As an aside, we are not reverting back to the version 1.x method of supporting short names as that was a hack and encouraged bad behavior which would result in things breaking when they updated. Then they would complain to us. The new way requires a proper installation, which avoids those issues.

[1]: https://python-markdown.github.io/reference/#extensions
